depthmapX/mainwindowpluginfactory.hpp

```diff
@@ -0,0 +1,52 @@
+// Copyright (C) 2020, Petros Koutsolampros
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Original found here: https://stackoverflow.com/a/582456
+
+#pragma once
+
+#include "depthmapX/mainwindowplugin.h"
+
+#include <iostream>
+
+template <typename T> MainWindowPlugin *createT() { return new T; }
+
+struct MainWindowPluginFactory {
```

Comment:
I do not like this class - you pretty much just hand out a static map for everyone to do with what they like!
Possible solutions (in order decreasing favourability):
- The plug-in registry as a standard class whose lifetime gets controlled by the respective main functions references to which get passed around to plug-in registration/where the plug-ins are needed.
- The plug-in registry as a global singleton that still encapsulates all the plug-in registration/finding logic, but only one (static) instance can exist
- This.